Revamp reading plan 2019-03-24

timbze opened the pull request:
Reading plan updates. Progress in #284 
Now using DB instead of asset files and sharedPrefs.

NEED some more testing before merging to master. One issue may be the time it takes to create DB tables and add current reading plan records the first time this version is loaded. There is a way to make it much faster, just needs to be done yet.

Closes #50
Closes #55 
Closes #292 
Closes #283 (need to update other translations)
Minor change toward #180 (There is now an "Import Plans" option in top right menu, which simply looks in SD jsword/readingplan folder for plans that are not yet in DB. Need some more changes)

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -50,162 +71,311 @@ import javax.inject.Inject
+            try {
+                readingPlanOneDay = ReadingPlanOneDayDB(readingPlanInfo)
+            } finally { }
+        } catch (e: Exception) {
```
In this specific location I should have done it differently anyway. I've now updated the code.

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
+
+    private fun speakAllReadings(allReadings: List<Key>) {
+        val bible = currentPageManager.currentBible.currentPassageBook
+        val allReadingsWithCorrectV11n = java.util.ArrayList<Key>()
```
```suggestion
        val allReadingsWithCorrectV11n: ArrayList<Key> = arrayListOf()

timbze replied:
Done

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/readingplan/ReadingPlanItemAdapter.kt:
```diff
@@ -18,45 +18,43 @@
  */
-class ReadingPlanItemAdapter(_context: Context, private val resource: Int, _items: List<ReadingPlanInfoDto>):
-		ArrayAdapter<ReadingPlanInfoDto>(_context, resource, _items) {
+class ReadingPlanItemAdapter(_context: Context, private val resource: Int, _items: List<ReadingPlanInformationDB>):
```
```suggestion
class ReadingPlanItemAdapter(context: Context, private val resource: Int, items: List<ReadingPlanInformationDB>):
```

Or do you have a reason for these _ prefixes?

timbze replied:
No reason. They were there from before and I haven't thought of changing. Will do it.

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/readingplan/actionbar/ReadingPlanActionBarManager.kt:
```diff
@@ -36,7 +36,7 @@ import javax.inject.Inject
 @ApplicationScope
 class ReadingPlanActionBarManager @Inject
-constructor(private val readingPlanTitle: ReadingPlanTitle, private val pauseActionBarButton: ReadingPlanPauseActionBarButton, private val stopActionBarButton: ReadingPlanStopActionBarButton, private val bibleActionBarButton: ReadingPlanBibleActionBarButton, private val commentaryActionBarButton: ReadingPlanCommentaryActionBarButton, private val dictionaryActionBarButton: ReadingPlanDictionaryAct ...
+constructor(val readingPlanTitle: ReadingPlanTitle, private val pauseActionBarButton: ReadingPlanPauseActionBarButton, private val stopActionBarButton: ReadingPlanStopActionBarButton, private val bibleActionBarButton: ReadingPlanBibleActionBarButton, private val commentaryActionBarButton: ReadingPlanCommentaryActionBarButton, private val dictionaryActionBarButton: ReadingPlanDictionaryActionBarBu ...
```
Please wrap this long line

timbze replied:
Done

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/readingplan/actionbar/ReadingPlanTitle.kt:
```diff
@@ -52,17 +51,18 @@ import javax.inject.Inject
     private lateinit var actionBar: ActionBar
     private lateinit var activity: Activity
-        private set
+    fun isActivityInitialized() = ::activity.isInitialized
```
I like property syntax more:

```suggestion
    val isActivityInitialized get() = ::activity.isInitialized
```

tuomas2 replied:
though, why not use ::activity.isInitialized directly where needed?

timbze replied:
I needed it in DailyReading.kt, and for some reason Kotlin does not allow this type of code to be used from another class.
I've changed it to property. Let me know if there's a better way of doing it.

tuomas2 replied:
Okay

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -50,162 +71,311 @@ import javax.inject.Inject
+                readingPlanOneDay = ReadingPlanOneDayDB(readingPlanInfo)
+            } finally { }
+        } catch (e: Exception) {
+            Log.e(TAG,"No current active reading plan. Will load plan selection screen.")
```
Normal condition. Updated the code.

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -50,162 +71,311 @@ import javax.inject.Inject
+            readingPlanInfo = ReadingPlanInformationDB(mReadingPlanMetaID)
+
+            try {
+                readingPlanOneDay = ReadingPlanOneDayDB(readingPlanInfo)
```
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -50,162 +71,311 @@ import javax.inject.Inject
+
+        val keyList = java.util.ArrayList<Key>()
+        keyList.add(convertedPassage)
+        return keyList
```
Done.

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -350,8 +582,51 @@ class DailyReading : CustomTitlebarActivityBase(R.menu.reading_plan) {
+        const val REQUEST_CODE_READING_PLAN_LIST = 101
+
+        // Link AB distributed reading plan file names with plan name/description resource strings
+        val ABDistributedPlanDetailArray: Array<PlanDetails> = arrayOf(
```
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -50,162 +71,311 @@ import javax.inject.Inject
+    /** User wants all passages from the daily reading spoken using TTS
+     * Also mark passages as read
+     */
+    fun speak(day: Int, allReadings: List<Key>) {
```
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
+
+        if (dbAdapter.metaCurrentActiveReadingPlanID == null) {
+            // Load plan selection screen
+            this.startActivityForResult(
```
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,725 @@
+    val readingPlanStartDate: Date? = dbAdapter.getMetaReadingPlanStartDate(readingPlanMetaID)
+    val readingPlanTotalDays: Int = dbAdapter.getMetaTotalDays(readingPlanMetaID)
+    val readingPlanVersificationName: String? = dbAdapter.getMetaVersificationName(readingPlanMetaID)
+    val readingPlanIsCustomUserAdded: Boolean = dbAdapter.getMetaIsCustomUserAddedPlan(readingPlanFileName)
```
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,725 @@
+/**
+ *
+ * @throws Exception when there is no [ReadingPlanMetaIdParam] provided AND ALSO can not get the current
+ * active reading plan from DB. It usually simply means that no plans have been started as yet
```
From ReadingPlanDBAdapter.kt line 519-523:
`private val dbAdapterStatic = ReadingPlanDBAdapter`

`val metaID: Int = ReadingPlanMetaIdParam ?: dbAdapter.metaCurrentActiveReadingPlanID!!`
`val isDateBased: Boolean = dbAdapter.getMetaIsDateBasedPlan(metaID)`
`val fileName: String = dbAdapter.getMetaFileNameFromId(metaID)`

If there is no active reading plan yet (i.e. first time loading), dbAdapter.metaCurrentActiveReadingPlanID will be null and throw error. The class should not be loaded if there is no metaId specified. Maybe there would be another better way of doing it? Earlier I used this in DailyReading.kt to try for error, and if there was an error then I knew the plan selector list needs to be loaded, but I changed that to just check if the metaCurrentActiveReadingPlanID is null and then load the selector list.

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,725 @@
+ * @throws Exception when there is no [ReadingPlanMetaIdParam] provided AND ALSO can not get the current
+ * active reading plan from DB. It usually simply means that no plans have been started as yet
+ */
+class ReadingPlanInformationDB(private val ReadingPlanMetaIdParam: Int?) {
```
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
+        forceReload: Boolean = false
+        ) {
+
+        val layout: TableLayout = findViewById(R.id.reading_container)
```
Some of the code is here from before and I haven't gone through all these type of details. But I appreciate your comments.
Fixed.

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
+    }
+
+    private fun clearReadingInfo(layout: TableLayout) {
+        layout.removeAllViews()
```
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
+     */
+    private fun loadPlanOneDay(
+        readingPlanOneDayParam: ReadingPlanOneDayDB?,
+        readingPlanInformationParam: ReadingPlanInformationDB? = null,
```
The reason I put the Param suffix at a few places is because I needed to use another variable throughout the function. In this case line 150, might be 154 for you:
`readingPlanOneDay.readingPlanInfo = readingPlanInformationParam`

In this case I'm setting the readingPlanInfo as readingPlanInformationParam. I'm trying to keep them distinguished and not have the same names.

tuomas2 replied:
Okay. Let's keep this like this for a while. Maybe I'll comment this later again. Though somehow I don't like it but don't know what to suggest either. 

Some options.
- You could have same name, and if you need to refer to the attribute within same context, you would use `this.readingPlanInformation`
- You could use `_` suffix (i.e. `readingPlanInfo_`). I think I have used that sometimes. (though I must admit that this convention comes from Python world)

timbze replied:
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
+            clearReadingInfo(layout)
+            description.text = readingPlanInformationParam.readingPlanName
+            date.text = SimpleDateFormat.getDateInstance().format(Calendar.getInstance().time)
+            status_message.text = app.getString(R.string.this_day_has_no_readings)
```
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
-            updateTicksAndDone()
+            updateTicksAndDone(readingPlanOneDay)
 
             // Speak All
```
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
+        date.text = readingPlanOneDay.dateString
+
+        if (readingPlanActionBarManager.readingPlanTitle.isActivityInitialized()) {
+            readingPlanActionBarManager.readingPlanTitle.update(StringUtils.left(readingPlanInfo.readingPlanFileName, 8), readingPlanOneDay.dayAndNumberDescription)
```
Done

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
-        if (mDay < mReadings.readingPlanInfo.numberOfPlanDays) {
-            showDay(mDay + 1)
+    private fun readOnePassage(readingNumber: Int, readingKey: Key?) {
+        if (readingKey != null) {
```
Good one. Done.

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
-        Log.i(TAG, "Previous")
-        if (mDay > 1) {
-            showDay(mDay - 1)
+    private fun convertReadingVersification(readingKey: Key, bibleToBeUsed: AbstractPassageBook): List<Key> {
```
I am not sure why. This is the code from before. But I see that if I would change it I would probably also have to change a lot more stuff, as SpeakControl.kt speakKeyList requires a List<Key> for parameter.

tuomas2 replied:
Okay, so let's leave this for now.

timbze commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -29,282 +30,431 @@ import android.widget.Button
-        readingPlanControl.read(mDay, readingNo, readingKey)
+    private fun setupOneReading(readingNumber: Int) {
+        val child = layoutInflater.inflate(R.layout.reading_plan_one_reading, null)
+        val imageTick = child.findViewById<View>(R.id.tick) as ImageView
```
Done

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,767 @@
+        }
+
+    val getCurrentPlanShortCode: String =
+        StringUtils.left(getMetaFileNameFromId(metaCurrentActiveReadingPlanID ?: 0),8)
```
It looks like you are trying to make a property here too. Now this is constant value (initialized when instance is created). Whether constant or property, please omit get prefix.

timbze replied:
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,767 @@
+        val app = BibleApplication.application
+        val dbHelper: CommonDatabaseHelper = CommonDatabaseHelper.getInstance()
+
+        private val dbWritable = dbHelper.writableDatabase
```
dbHelper has a writable and readable DB object. One is Read/Write, other is read-only. I think in this place it only needs to be set at initialization. Maybe I'm wrong.

tuomas2 replied:
Okay, I see. getReadableDatabase seems to give writable database as well if possible, it just does not fail like getWritableDatabase would if it is not possible to open it in read-write mode (so I understood from the documentation of these functions). So to me it looks like only one would be enough. Though I don't know best conventions here.

timbze replied:
Ok I didn't know that readableDatabase also gets read/write. I will just use that.

Is it better to use it as a property? I'm just wondering cause the documentation says it could take long to load (even though for such small db file it shouldn't be too long). I'm not sure if it's good to constantly be loading it then.

tuomas2 replied:
I think constant indeed is correct way after all.

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,767 @@
+
+    var metaCurrentActiveReadingPlanID: Int?
+        get() {
+            val projection = arrayOf(BaseColumns._ID)
```
Well that's what Sqlite function calls it. Search for projection on this page: https://developer.android.com/training/data-storage/sqlite

If you still think it's better to use another name then we could do that.

tuomas2 replied:
Ah, I see. API (android.database.sqlite.SQLiteDatabase#query) calls that parameter `columns` which makes more sense to me.

timbze replied:
I guess I should maybe just use columns instead anyway. Easier to understand for everyone.

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,767 @@
+                cursor.close()
+                result = readID
+            }
+            cursor.close()
```
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,767 @@
+            cursor.close()
+            return result
+        }
+        set(ReadPlanID) {
```
Done

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,770 @@
+        private val dbReadable = dbHelper.readableDatabase
+
+        private val dateBasedFormatMonthDay = SimpleDateFormat("MMM-d")
+        val dateBasedFormatWithYear = SimpleDateFormat("MMM-d/yyyy")
```
Maybe all of these could be private? It does not make sense that they would be used outside of this class?

timbze replied:
I changed a few. dateBasedFormatWithYear I need to use in another place too.

tuomas2 replied:
If they are used in multiple classes and are not strictly semantically connected to the particular class, they might be module level variables too.

timbze replied:
I thought I had commented here but now here is no comment, so here it is again. Because of the way I am requesting the Dates in date-based text files to be (i.e. Feb-1, Mar-22, etc), I need to be using the exact format that is specified on dateBasedFormatWithYear. And I am also using it in another class. I think it's a better idea to just specify the format in one place than to have it in more places.

tuomas2 replied:
I agree but what is the best place, that's sometimes difficult question. Could be some utils file/class. I was pointing out that these do not need to be in class (companion object) level, but on module level too. It means that you would have these lines in the beginning of a .kt file. 

But BTW now I can't find dateBasedFormatWithYear any more outside of this class.

timbze replied:
Done

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,770 @@
+                selectionArgs,
+                null,
+                null,
+                null)
```
You could consider putting those arguments inline without creating separate variables

timbze replied:
Done

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,770 @@
+                val setFalseValues = ContentValues().apply {
+                    put(ReadingPlanMeta.COLUMN_LAST_USED_PLAN, DB_FALSE_VALUE)
+                }
+                val setTheTrueValue = ContentValues().apply {
```
inconsistency in variable naming. Maybe

```suggestion
                val setTrueValues = ContentValues().apply {
```

timbze replied:
Well my reason for the name is: setFalseValues (first mark every row as false). setTheTrueValue (Set the ONE reading plan as true which was last used)

tuomas2 replied:
Okay, I see...

tuomas2 replied:
Essentially you are wanting to store which reading plan is active, right? Then I think you should write somewhere just the index of that entry, instead of boolean value on each column.

tuomas2 replied:
If you change it like that, this property will be a lot simpler and cleaner code. I'm thinking that most likely you could also make this non-nullable (with a sane default value, like 0).

timbze replied:
I think that's a good idea. Fixed.

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,770 @@
+
+    }
+
+    var metaCurrentActiveReadingPlanID: Int?
```
What do you mean by calling this with meta prefix?

timbze replied:
As per our discussion, have dropped most "meta" strings from name.

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,770 @@
+            getMetaCurrentDayNumber(metaCurrentActiveReadingPlanID ?: 0).toString()
+        )
+    /** Will always get or set the current day number from the active reading plan
+     * @return Null result means that it is a date-based plan and there are no readings for today.
```
this can't return null

timbze replied:
Duh, I guess that's a comment from when I still did return null :)
Fixed!

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,770 @@
+            return when (currentDay) {
+                0 -> 1
+                else -> currentDay
+            }
```
perhaps `if` rather than `when` because there are only 2 options

timbze replied:
Done

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,770 @@
+     * from DB with second of this function with double constructor, here: [getMetaReadingPlanDescription]
+     */
+    fun getMetaReadingPlanDescription(ReadingPlanMetaID: Int): String {
+        return getPlanDetailLink(ReadingPlanMetaID)?.planDescription ?:
```
weird name getPlanDetailLink -- what do you mean by 'link'?

timbze replied:
I guess it was just my way of easily distinguishing from DB data. With just PlanDetail it wouldn't be clear right away that this is not info from DB. With the link in the name, to me it's the link between DB and resource strings. But anyway, don't know if it's good or not.

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,770 @@
+    /** Will always get or set the current day number from the active reading plan
+     * @return Null result means that it is a date-based plan and there are no readings for today.
+     */
+    fun getMetaCurrentDayNumber(readingPlanMetaId: Int): Int {
```
As discussed, please consider dropping Meta from all methods and variable / parameter names (you may leave the table with that name though). Please think if the code is still understandable and nice after the change.

timbze replied:
Done. I left a few private functions with meta in the name, since that describes what they do.

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDatabaseDefinition.kt:
```diff
@@ -0,0 +1,246 @@
+                "${ReadingPlanMeta.COLUMN_VERSIFICATION_NAME} TEXT NOT NULL," +
+                "${ReadingPlanMeta.COLUMN_LAST_USED_PLAN} INTEGER NOT NULL DEFAULT 0" +
+                "); "
+
```
Please use Kotlin multiline string literals  (`"""..."""`)

tuomas2 replied:
Here and in other similar places

timbze replied:
> If they are used in multiple classes and are not strictly semantically connected to the particular class, they might be module level variables too.

Can you give me an example of what you mean with module level variable? I do not really want to put this line more than once, since it has to do with the specific way the date-based plan needs to have the date specified: SimpleDateFormat("MMM-d/yyyy") -- I am also using it in OneDaysReadingsDto.

timbze replied:
> Essentially you are wanting to store which reading plan is active, right? Then I think you should write somewhere just the index of that entry, instead of boolean value on each column.

Yes, would you make another table just for that? I don't see how I could do that in the 2 tables I'm using now.

tuomas2 replied:
Until then maybe this could be stored in current shared preferences.

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/readingplan/DailyReading.kt:
```diff
@@ -313,29 +464,56 @@ class DailyReading : CustomTitlebarActivityBase(R.menu.reading_plan) {
-                day.text = mReadings.dayDesc
+                isHandled = true
+            }
+            R.id.importNewPlan -> {
```
Manual importing new plans should be done via ACTION_GET_CONTENT intent (similarly as in InstallZip). Then this does not need any permissions at all and user can choose importable file from any source (like google drive).

timbze replied:
Done

tuomas2 commented on app/src/main/java/net/bible/service/readingplan/ReadingPlanDao.kt:
```diff
@@ -65,6 +68,25 @@ class ReadingPlanDao {
+        get() {
+            try {
+                val codes = sdReadingPlanCodes
+                if (codes == null) { return null }
```
```kt
val codes = sdReadingPlanCodes ?: return null
```

timbze replied:
Removed the class altogether

tuomas2 commented on app/src/main/java/net/bible/service/readingplan/ReadingPlanDao.kt:
```diff
@@ -65,6 +68,25 @@ class ReadingPlanDao {
 
         }
 
+    val sdReadingPlanList: List<ReadingPlanInfoDto>?
```
sd? Modern Android device does not necessarily have SD card, so probably we shouldn't use it any more in our terminology (although for legacy reasons it has been used in the codebase).

What is the correct term then? User storage? (vs. app storage, which user can't access but app can) Not sure. You could check this from documentation.

timbze replied:
No longer applies. Class deleted

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDatabaseDefinition.kt:
```diff
@@ -0,0 +1,246 @@
+                        Log.d(TAG, "There are no reading plans on SD card.")
+                        return
+                    }
+                    readingPlanCodes = ReadingPlanDao().sdReadingPlanList!!
```
Lines 124-128
```kt
                    val sdReadingPlanList = ReadingPlanDao().sdReadingPlanList
                    if (sdReadingPlanList == null) {
                        Log.d(TAG, "There are no reading plans on SD card.")
                        return
                    }
                    readingPlanCodes = sdReadingPlanList
```

timbze replied:
Changed. Doesn't apply anymore.

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDatabaseDefinition.kt:
```diff
@@ -0,0 +1,246 @@
+                } else {
+                    // Get all Reading Plans list
+                    readingPlanCodes = ReadingPlanDao().readingPlanList
+                }
```
120-132
```kt
                val readingPlanDao = ReadingPlanDao()
                if (!firstImport) {
                    // Get SD reading plans list
                    dbAdapter = ReadingPlanDBAdapter()
                    val sdReadingPlanList = readingPlanDao.sdReadingPlanList
                    if (sdReadingPlanList == null) {
                        Log.d(TAG, "There are no reading plans on SD card.")
                        return
                    }
                    readingPlanCodes = sdReadingPlanList
                } else {
                    // Get all Reading Plans list
                    readingPlanCodes = readingPlanDao.readingPlanList
                }
```

timbze replied:
Changed code. Doesn't apply anymore

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDatabaseDefinition.kt:
```diff
@@ -0,0 +1,246 @@
+
+        }
+
+    }
```
again mega-long function. Please refactor into smaller functions (multiple new private helper functions).

timbze replied:
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,750 @@
+        private val app = BibleApplication.application
+        private val dbHelper: CommonDatabaseHelper = CommonDatabaseHelper.getInstance()
+
+        private val dbReadable = dbHelper.readableDatabase
```
Done

timbze commented:
I have now added commits for changes relating to custom user plan import. It was my plan from the beginning to do this, but now I have removed 3 of the remaining "old" files that I just used pieces of, and adapted that code into ReadingPlanDatabaseDefinition (I think at this point it might be better to split that class into another class, let me know what you think).

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDatabaseDefinition.kt:
```diff
@@ -0,0 +1,499 @@
+                ${ReadingPlanMeta.COLUMN_CURRENT_DAY} INTEGER NOT NULL DEFAULT 0,
+                ${ReadingPlanMeta.COLUMN_VERSIFICATION_NAME} TEXT NOT NULL
+                ); """
+
```
I was experimenting with kotlin a little and found out that you could write this a bit neater like this:

```kt
            private val SQL_CREATE_ENTRIES_META = ReadingPlanMeta.run {
                """CREATE TABLE ${ReadingPlanMeta.TABLE_NAME} (
                ${COLUMN_ID} INTEGER PRIMARY KEY AUTOINCREMENT,
                ${COLUMN_PLAN_FILE_NAME} TEXT NOT NULL,
                ${COLUMN_PLAN_NAME} TEXT NOT NULL,
                ${COLUMN_DESCRIPTION} TEXT NOT NULL,
                ${COLUMN_DATE_START} INTEGER,
                ${COLUMN_DAYS_IN_PLAN} INTEGER,
                ${COLUMN_CURRENT_DAY} INTEGER NOT NULL DEFAULT 0,
                ${COLUMN_VERSIFICATION_NAME} TEXT NOT NULL
                ); """
            }
```

tuomas2 replied:
If you like this, change all others similarly as well.

timbze replied:
Nice. Done

tuomas2 commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDatabaseDefinition.kt:
```diff
@@ -0,0 +1,499 @@
+                var returnValue = false
+
+                readingPlanProperties = when (firstImport) {
+                    true -> getAllReadingPlanProperties
```
```suggestion
                    true -> allReadingPlanProperties
```

(property naming convention)

timbze replied:
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,769 @@
+        } else {
+
+            val currentDay: Int = getMetaIntegerDB(readingPlanMetaId, ReadingPlanMeta.COLUMN_CURRENT_DAY) ?: 1
+            return if (currentDay == 0) 1
```
With max it would work, not min :)
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,769 @@
+        app.getString(
+            R.string.rdg_plan_day,
+            getCurrentDayNumber(currentActiveReadingPlanID).toString()
+        )
```
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,769 @@
+        private val app = BibleApplication.application
+        private val dbHelper: CommonDatabaseHelper = CommonDatabaseHelper.getInstance()
+
+        private val writableDatabase = dbHelper.readableDatabase
```
Done

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDBAdapter.kt:
```diff
@@ -0,0 +1,769 @@
+        return result
+    }
+
+    private fun getMetaIntegerDB(
```
That generic function was a great idea. Didn't know that was possible. Used 1 generic now instead of about 4 other functions.

timbze commented:
I see PR #124 also has a DB upgrade, so whichever PR is merged last, this or that one, will have to change DB version again, since this PR is also upgrading DB.

tuomas2 commented:
> I see PR #124 also has a DB upgrade, so whichever PR is merged last, this or that one, will have to change DB version again, since this PR is also upgrading DB.

Author of that PR has not been active lately. I'll probably close that PR at some point as it is getting old.

timbze commented on app/src/main/java/net/bible/service/db/readingplan/ReadingPlanDatabaseDefinition.kt:
```diff
@@ -0,0 +1,499 @@
+    object ReadingPlanDays : BaseColumns {
+        const val TABLE_NAME = "readingplan_days"
+        const val COLUMN_ID = BaseColumns._ID
+        const val COLUMN_READING_PLAN_META_ID = "readingplan_meta_id"
```
Think that's good idea. Done

tuomas2 commented:
Is there any easy way to go forward 1 day or go back 1 day? Browsing days should be simple(r). With buttons probably the clearest, but in addition we could also support swipe gesture?

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/readingplan/ReadingPlanSelectorList.kt:
```diff
@@ -73,23 +70,27 @@ class ReadingPlanSelectorList : ListActivityBase() {
         } catch (e: Exception) {
             Log.e(TAG, "Plan selection error", e)
             Dialogs.getInstance().showErrorMsg(R.string.error_occurred, e)
         }
```
(Again) I would like us to re-consider if this is such a place in our code, where we are expecting errors to take place easily (because of user mistake for example -- for example error in reading plan document)

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/readingplan/actionbar/ReadingPlanTitle.kt:
```diff
@@ -31,12 +31,11 @@ import android.widget.TextView
 
 import net.bible.android.activity.R
```
What plans do you have for this action bar? Planning to keep it like this for now?

timbze commented:
Closing this as we are now doing this in smaller steps without the major db change.
